File: src/bin/alt_deploy.rs
//! Deploy an address-lookup-table (ALT) on-chain from a list of accounts on
//! stdin — so the box doesn't need the `solana` CLI to create SAVE_ALT / JUP_ALT.
//!
//! Reads pubkeys (one per line) from stdin, creates a fresh lookup table owned by
//! the KEYPAIR_PATH wallet, extends it with those addresses (batched to fit the
//! 1232B tx limit), and prints the resulting table address.
//!
//! Pipe it from the *_alt_print bins:
//!   LIVE=1 cargo run --release --bin save_alt_print | cargo run --release --bin alt_deploy
//!   LIVE=1 cargo run --release --bin jup_alt_print  | cargo run --release --bin alt_deploy
//! then `export SAVE_ALT=<printed table>` (or JUP_ALT=…).
//!
//! SAFETY: DRY-RUN by default — it only prints the plan. Set LIVE=1 to submit
//! real txs (creates on-chain state + spends a little SOL for rent + fees, signed
//! by KEYPAIR_PATH). Uses the official solana-address-lookup-table-interface
//! instruction builders, not a hand-rolled format.
//!
//! Env: HELIUS_RPC|RPC_HTTP|RPC_ENDPOINT, KEYPAIR_PATH, [LIVE=1], [CU_PRICE=<micro-lamports>]

use arb_engine::arb::{cu_limit_ix, cu_price_ix};
use base64::Engine;
use solana_address_lookup_table_interface::instruction::{create_lookup_table, extend_lookup_table};
use solana_hash::Hash;
use solana_keypair::Keypair;
use solana_message::{v0, VersionedMessage};
use solana_pubkey::Pubkey;
use solana_signer::Signer;
use solana_transaction::versioned::VersionedTransaction;
use std::io::Read;
use std::str::FromStr;
use std::time::{Duration, Instant};

/// Addresses per extend tx. Each pubkey is 32B of ix data; 20 keeps the tx well
/// under the 1232B single-packet limit with room for the header + signature.
const EXTEND_BATCH: usize = 20;

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(r) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = r.into_json::<serde_json::Value>() {
                return Some(v);
            }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}

fn latest_blockhash(endpoint: &str) -> Hash {
    let v = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getLatestBlockhash",
        "params":[{"commitment":"finalized"}]})).expect("getLatestBlockhash");
    Hash::from_str(v["result"]["value"]["blockhash"].as_str().expect("blockhash")).unwrap()
}

/// Build a v0 tx from `ixs`, sign with `kp`, submit, and poll until confirmed.
/// Returns the signature. Exits the process on rejection or confirm timeout.
fn send_and_confirm(
    endpoint: &str,
    kp: &Keypair,
    ixs: &[solana_instruction::Instruction],
    label: &str,
) -> String {
    let bh = latest_blockhash(endpoint);
    let msg = v0::Message::try_compile(&kp.pubkey(), ixs, &[], bh).expect("compile");
    let mut tx = VersionedTransaction {
        signatures: vec![solana_signature::Signature::default()],
        message: VersionedMessage::V0(msg),
    };
    tx.signatures[0] = kp.sign_message(&tx.message.serialize());
    let sig = tx.signatures[0].to_string();
    let b64 = base64::engine::general_purpose::STANDARD.encode(bincode::serialize(&tx).unwrap());

    let v = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"sendTransaction",
        "params":[b64,{"encoding":"base64","skipPreflight":false,"preflightCommitment":"confirmed","maxRetries":5}]}))
        .expect("sendTransaction");
    if let Some(e) = v.get("error").filter(|e| !e.is_null()) {
        eprintln!("⛔ {label}: sendTransaction rejected: {e}");
        std::process::exit(1);
    }
    println!("  {label}: submitted {sig} — confirming…");

    let start = Instant::now();
    while start.elapsed() < Duration::from_secs(45) {
        std::thread::sleep(Duration::from_millis(1500));
        let s = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getSignatureStatuses",
            "params":[[sig], {"searchTransactionHistory":false}]}));
        let Some(s) = s else { continue };
        let st = &s["result"]["value"][0];
        if st.is_null() {
            continue;
        }
        if !st["err"].is_null() {
            eprintln!("⛔ {label}: tx failed on-chain: {}", st["err"]);
            std::process::exit(1);
        }
        let cs = st["confirmationStatus"].as_str().unwrap_or("");
        if cs == "confirmed" || cs == "finalized" {
            println!("  {label}: ✅ {cs}");
            return sig;
        }
    }
    eprintln!("⛔ {label}: not confirmed within 45s (sig {sig}); check explorer before retrying to avoid a duplicate table");
    std::process::exit(1);
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC")
        .or_else(|_| std::env::var("RPC_HTTP"))
        .or_else(|_| std::env::var("RPC_ENDPOINT"))
        .expect("set HELIUS_RPC / RPC_HTTP / RPC_ENDPOINT");
    let keypair_path = std::env::var("KEYPAIR_PATH").expect("set KEYPAIR_PATH");
    let live = std::env::var("LIVE").map(|v| v == "1").unwrap_or(false);
    let cu_price: u64 = std::env::var("CU_PRICE").ok().and_then(|s| s.parse().ok()).unwrap_or(100_000);

    let bytes: Vec<u8> = serde_json::from_str(&std::fs::read_to_string(&keypair_path).expect("read keypair"))
        .expect("parse keypair (expects a JSON byte array)");
    let kp = Keypair::try_from(&bytes[..]).expect("keypair");
    let authority = kp.pubkey();

    // Read addresses from stdin (one per line). Non-parseable lines are skipped
    // so piping a print bin's stdout (addresses only; notes go to stderr) is clean.
    let mut input = String::new();
    std::io::stdin().read_to_string(&mut input).expect("read stdin");
    let mut seen = std::collections::HashSet::new();
    let addrs: Vec<Pubkey> = input
        .lines()
        .filter_map(|l| Pubkey::from_str(l.trim()).ok())
        .filter(|p| seen.insert(*p))
        .collect();
    if addrs.is_empty() {
        eprintln!("⛔ no valid pubkeys on stdin — pipe from save_alt_print / jup_alt_print");
        std::process::exit(1);
    }

    // Recent (finalized) slot for the CreateLookupTable derivation.
    let slot = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getSlot",
        "params":[{"commitment":"finalized"}]})).expect("getSlot")["result"]
        .as_u64()
        .expect("slot");
    let (create_ix, table) = create_lookup_table(authority, authority, slot);
    let batches = addrs.len().div_ceil(EXTEND_BATCH);

    println!("ALT deploy plan:");
    println!("  authority/payer : {authority}");
    println!("  addresses       : {}", addrs.len());
    println!("  recent slot     : {slot}");
    println!("  table address   : {table}");
    println!("  txs             : 1 create + {batches} extend (batches of {EXTEND_BATCH})");

    if !live {
        println!("\nDRY RUN — nothing submitted. Set LIVE=1 to deploy for real.");
        println!("NOTE: the table address above is derived from the current slot and will DIFFER on the live run;");
        println!("      the real address is printed at the end of the LIVE run.");
        return;
    }

    println!("\nLIVE — submitting…");
    send_and_confirm(&endpoint, &kp, &[cu_limit_ix(60_000), cu_price_ix(cu_price), create_ix], "create");
    for (i, chunk) in addrs.chunks(EXTEND_BATCH).enumerate() {
        let ix = extend_lookup_table(table, authority, Some(authority), chunk.to_vec());
        send_and_confirm(&endpoint, &kp, &[cu_limit_ix(60_000), cu_price_ix(cu_price), ix], &format!("extend {}/{batches}", i + 1));
    }

    println!("\n✅ ALT deployed with {} addresses.", addrs.len());
    println!("   table = {table}");
    println!("   → export the matching var, e.g.  export SAVE_ALT={table}   (or JUP_ALT={table})");
    println!("   (an ALT needs ~1 slot to warm up before it can be used in a tx.)");
}

File: src/bin/altDeploy.ts
// Port of src/bin/alt_deploy.rs
//
// Deploy an address-lookup-table (ALT) on-chain from a list of accounts on
// stdin — so the box doesn't need the `solana` CLI to create SAVE_ALT / JUP_ALT.
//
// Reads pubkeys (one per line) from stdin, creates a fresh lookup table owned by
// the KEYPAIR_PATH wallet, extends it with those addresses (batched to fit the
// 1232B tx limit), and prints the resulting table address.
//
// Pipe it from the *_alt_print bins:
//   LIVE=1 tsx src/bin/saveAltPrint.ts | tsx src/bin/altDeploy.ts
//   LIVE=1 tsx src/bin/jupAltPrint.ts  | tsx src/bin/altDeploy.ts
// then `export SAVE_ALT=<printed table>` (or JUP_ALT=…).
//
// SAFETY: DRY-RUN by default — it only prints the plan. Set LIVE=1 to submit
// real txs (creates on-chain state + spends a little SOL for rent + fees, signed
// by KEYPAIR_PATH). Uses the official solana-address-lookup-table-interface
// instruction builders, not a hand-rolled format.
//
// Env: HELIUS_RPC|RPC_HTTP|RPC_ENDPOINT, KEYPAIR_PATH, [LIVE=1], [CU_PRICE=<micro-lamports>]

import 'dotenv/config';
import { readFileSync } from 'node:fs';
import readline from 'node:readline';
import {
  AddressLookupTableProgram,
  ComputeBudgetProgram,
  Keypair,
  PublicKey,
  TransactionMessage,
  VersionedTransaction,
} from '@solana/web3.js';

/** Addresses per extend tx. Each pubkey is 32B of ix data; 20 keeps the tx well
 * under the 1232B single-packet limit with room for the header + signature. */
const EXTEND_BATCH = 20;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await res.json();
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

async function latestBlockhash(endpoint: string): Promise<string> {
  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getLatestBlockhash',
    params: [{ commitment: 'finalized' }],
  });
  const bh = v?.result?.value?.blockhash;
  if (typeof bh !== 'string') throw new Error('getLatestBlockhash');
  return bh;
}

function cuLimitIx(units: number) {
  return ComputeBudgetProgram.setComputeUnitLimit({ units });
}
function cuPriceIx(microLamports: number) {
  return ComputeBudgetProgram.setComputeUnitPrice({ microLamports });
}

/**
 * Build a v0 tx from `ixs`, sign with `kp`, submit, and poll until confirmed.
 * Returns the signature. Exits the process on rejection or confirm timeout.
 */
async function sendAndConfirm(endpoint: string, kp: Keypair, ixs: any[], label: string): Promise<string> {
  const bh = await latestBlockhash(endpoint);
  const msg = new TransactionMessage({
    payerKey: kp.publicKey,
    recentBlockhash: bh,
    instructions: ixs,
  }).compileToV0Message();
  const tx = new VersionedTransaction(msg);
  tx.sign([kp]);
  const b64 = Buffer.from(tx.serialize()).toString('base64');

  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'sendTransaction',
    params: [b64, { encoding: 'base64', skipPreflight: false, preflightCommitment: 'confirmed', maxRetries: 5 }],
  });
  if (v === undefined) throw new Error('sendTransaction');
  if (v?.error != null) {
    console.error(`⛔ ${label}: sendTransaction rejected: ${JSON.stringify(v.error)}`);
    process.exit(1);
  }
  const txSig: string = v.result;
  console.log(`  ${label}: submitted ${txSig} — confirming…`);

  const start = Date.now();
  while (Date.now() - start < 45_000) {
    await sleep(1500);
    const s = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getSignatureStatuses',
      params: [[txSig], { searchTransactionHistory: false }],
    });
    if (s === undefined) continue;
    const st = s?.result?.value?.[0];
    if (st === null || st === undefined) continue;
    if (st?.err !== null && st?.err !== undefined) {
      console.error(`⛔ ${label}: tx failed on-chain: ${JSON.stringify(st.err)}`);
      process.exit(1);
    }
    const cs = st?.confirmationStatus ?? '';
    if (cs === 'confirmed' || cs === 'finalized') {
      console.log(`  ${label}: ✅ ${cs}`);
      return txSig;
    }
  }
  console.error(`⛔ ${label}: not confirmed within 45s (sig ${txSig}); check explorer before retrying to avoid a duplicate table`);
  process.exit(1);
}

async function readStdin(): Promise<string> {
  const rl = readline.createInterface({ input: process.stdin, terminal: false });
  const lines: string[] = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines.join('\n');
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP ?? process.env.RPC_ENDPOINT;
  if (endpoint === undefined) throw new Error('set HELIUS_RPC / RPC_HTTP / RPC_ENDPOINT');
  const keypairPath = process.env.KEYPAIR_PATH;
  if (keypairPath === undefined) throw new Error('set KEYPAIR_PATH');
  const live = process.env.LIVE === '1';
  const cuPrice = Number.parseInt(process.env.CU_PRICE ?? '', 10) || 100_000;

  const bytes: number[] = JSON.parse(readFileSync(keypairPath, 'utf8'));
  const kp = Keypair.fromSecretKey(new Uint8Array(bytes));
  const authority = kp.publicKey;

  // Read addresses from stdin (one per line). Non-parseable lines are skipped
  // so piping a print bin's stdout (addresses only; notes go to stderr) is clean.
  const input = await readStdin();
  const seen = new Set<string>();
  const addrs: PublicKey[] = [];
  for (const l of input.split('\n')) {
    const trimmed = l.trim();
    if (trimmed.length === 0) continue;
    let pk: PublicKey;
    try {
      pk = new PublicKey(trimmed);
    } catch {
      continue;
    }
    const s = pk.toBase58();
    if (seen.has(s)) continue;
    seen.add(s);
    addrs.push(pk);
  }
  if (addrs.length === 0) {
    console.error('⛔ no valid pubkeys on stdin — pipe from save_alt_print / jup_alt_print');
    process.exit(1);
  }

  // Recent (finalized) slot for the CreateLookupTable derivation.
  const slotV = await rpc(endpoint, { jsonrpc: '2.0', id: 1, method: 'getSlot', params: [{ commitment: 'finalized' }] });
  const slot: number | undefined = slotV?.result;
  if (slot === undefined) throw new Error('getSlot');
  const [createIx, table] = AddressLookupTableProgram.createLookupTable({
    authority,
    payer: authority,
    recentSlot: slot,
  });
  const batches = Math.ceil(addrs.length / EXTEND_BATCH);

  console.log('ALT deploy plan:');
  console.log(`  authority/payer : ${authority.toBase58()}`);
  console.log(`  addresses       : ${addrs.length}`);
  console.log(`  recent slot     : ${slot}`);
  console.log(`  table address   : ${table.toBase58()}`);
  console.log(`  txs             : 1 create + ${batches} extend (batches of ${EXTEND_BATCH})`);

  if (!live) {
    console.log('\nDRY RUN — nothing submitted. Set LIVE=1 to deploy for real.');
    console.log('NOTE: the table address above is derived from the current slot and will DIFFER on the live run;');
    console.log('      the real address is printed at the end of the LIVE run.');
    return;
  }

  console.log('\nLIVE — submitting…');
  await sendAndConfirm(endpoint, kp, [cuLimitIx(60_000), cuPriceIx(cuPrice), createIx], 'create');
  for (let i = 0; i * EXTEND_BATCH < addrs.length; i++) {
    const chunk = addrs.slice(i * EXTEND_BATCH, (i + 1) * EXTEND_BATCH);
    const ix = AddressLookupTableProgram.extendLookupTable({
      payer: authority,
      authority,
      lookupTable: table,
      addresses: chunk,
    });
    await sendAndConfirm(endpoint, kp, [cuLimitIx(60_000), cuPriceIx(cuPrice), ix], `extend ${i + 1}/${batches}`);
  }

  console.log(`\n✅ ALT deployed with ${addrs.length} addresses.`);
  console.log(`   table = ${table.toBase58()}`);
  console.log(`   → export the matching var, e.g.  export SAVE_ALT=${table.toBase58()}   (or JUP_ALT=${table.toBase58()})`);
  console.log('   (an ALT needs ~1 slot to warm up before it can be used in a tx.)');
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
